src/phoenix/experimental/evals/retrievals.py

```diff
@@ -0,0 +1,82 @@
+"""
+Helper functions for evaluating the retrieval step of retrieval-augmented generation.
+"""
+
+from typing import List, Optional, Tuple
+
+import openai
+from tenacity import (
+    retry,
+    stop_after_attempt,
+    wait_random_exponential,
+)
+
+_EVALUATION_SYSTEM_MESSAGE = (
+    "You will be given a query and a reference text. "
+    "You must determine whether the reference text contains an answer to the input query. "
+    'Your response must be single word, either "relevant" or "irrelevant", '
+    "and should not contain any text or characters aside from that word. "
+    '"irrelevant" means that the reference text does not contain an answer to the query. '
+    '"relevant" means the reference text contains an answer to the query.'
+)
+_QUERY_CONTEXT_PROMPT_TEMPLATE = """# Query: {query}
+
+# Reference: {reference}
+
+# Answer ("relevant" or "irrelevant"): """
+
+
+def compute_precisions_at_k_and_relevance_classifications(
```

Comment:
We could simplify this function by having it compute only the relevance classification and only on one query at a time, as follows.

```python
def compute_relevance_classifications(
    query: str,
    retrieved_documents: List[str],
    model_name: str = "gpt-4",
) -> List[Optional[bool]]
```

then in the notebook we would do

```python
relevance_classifications = [compute_relevance_classifications(query, documents, evaluation_model_name) for query, documents in zip(query_texts, list_of_document_text_lists)]
list_of_precisions_at_k_lists = list(map(compute_precisions_at_k, relevance_classifications))
```

It's more natural to accept just the query paired with its list of retrieved documents. Having them aggregated in separate lists is a choice of the caller, and not something we need to accommodate, or should force onto the caller (if that's not what they have). But ultimately, this function is unnecessary, because it's just a double list comprehension.

```python
relevance_classifications = [[classify_relevance(query, document, evaluation_model_name) for document in documents] for query, documents in zip(query_texts, list_of_document_text_lists)]
```

Reply:
This is a good point. I removed the complicated function and just have two functions:
```
def compute_precisions_at_k(
    relevance_classifications: List[Optional[bool]],
) -> List[Optional[float]]:
```
and
```
def classify_relevance(query: str, document: str, model_name: str) -> Optional[bool]:
```